subprojects/resources-sftp/src/test/groovy/org/gradle/internal/resource/transport/sftp/SftpClientFactoryTest.groovy

```diff
@@ -255,6 +256,57 @@ class SftpClientFactoryTest extends ConcurrentSpec {
         actualClient1 != actualClient2
     }
 
+    def "creates a new client when no existing clients are connected"() {
+        def mockSftpClient1 = Mock(LockableSftpClient)
+        def mockSftpClient2 = Mock(LockableSftpClient)
+        def mockSftpClient3 = Mock(LockableSftpClient)
+
+        given:
+        URI uri = new URI('http://localhost:22/repo')
+        PasswordCredentials credentials = new DefaultPasswordCredentials('sftp', 'sftp')
+
+        when:
+        LockableSftpClient client1 = sftpClientFactory.createSftpClient(uri, credentials)
+        LockableSftpClient client2 = sftpClientFactory.createSftpClient(uri, credentials)
+        sftpClientFactory.releaseSftpClient(client1)
+        sftpClientFactory.releaseSftpClient(client2)
+        LockableSftpClient client3 = sftpClientFactory.createSftpClient(uri, credentials)
+
+        then:
+        3 * sftpClientCreator.createNewClient(new SftpHost(uri, credentials)) >>> [mockSftpClient1,mockSftpClient2,mockSftpClient3]
+        1 * mockSftpClient1.host >> new SftpHost(uri, credentials)
+        1 * mockSftpClient2.host >> new SftpHost(uri, credentials)
+        1 * mockSftpClient1.connected >> false
+        1 * mockSftpClient2.connected >> false
+        sftpClientFactory.clients.size() == 0
+        client3 == mockSftpClient3
+    }
+
+    def "reuses a different client when an existing client is no longer connected"() {
+        def mockSftpClient1 = Mock(LockableSftpClient)
+        def mockSftpClient2 = Mock(LockableSftpClient)
+
+        given:
+        URI uri = new URI('http://localhost:22/repo')
+        PasswordCredentials credentials = new DefaultPasswordCredentials('sftp', 'sftp')
+
+        when:
+        LockableSftpClient client1 = sftpClientFactory.createSftpClient(uri, credentials)
+        LockableSftpClient client2 = sftpClientFactory.createSftpClient(uri, credentials)
+        sftpClientFactory.releaseSftpClient(client1)
+        sftpClientFactory.releaseSftpClient(client2)
+        LockableSftpClient client3 = sftpClientFactory.createSftpClient(uri, credentials)
+
+        then:
+        2 * sftpClientCreator.createNewClient(new SftpHost(uri, credentials)) >>> [mockSftpClient1,mockSftpClient2]
+        1 * mockSftpClient1.host >> new SftpHost(uri, credentials)
+        1 * mockSftpClient2.host >> new SftpHost(uri, credentials)
+        1 * mockSftpClient1.connected >> false
+        1 * mockSftpClient2.connected >> true
+        sftpClientFactory.clients.size() == 0
```

Comment:
Because it was tracking idle clients and after we claim the one good client, there are no other idle clients left (i.e. we discard client1 because it's disconnected).  I've changed this to "idleClients" to be more clear.